src/representer.nim

```diff
@@ -0,0 +1,143 @@
+  ##   Empty | GenericParams
+  ##     IdentDefs
+  ##       Ident
+  ##       Type
+  ##   FormalParams
+  ##     empty | returntype
+  ##     @[IdentDefs] =
+  ##       idents | @[ident, ident]
+  ##       ident -> type
+  ##       empty | default
+  ##   Pragmas
+  ##   Empty
+  ##   StmtList # meat and potatoes
+  let formalParams = routineDef[3]
+  let returnType = formalParams[0]
+  let identDefs = formalParams[1..^1].mapIt(it.normalizeIdentDef(map))
+  routineDef.kind.newTree(
+    routineDef[0].addNewName(map), # RoutineDef Name
+    newEmptyNode(), # Term Rewriting macros and templates which are not supported
+    newEmptyNode(), # TODO: Implement Generic Params
+    nnkFormalParams.newTree(
+      returnType.normalizeValue(map)
+    ).add(identDefs),
+    newEmptyNode(), # TODO: Implement pragma
+    newEmptyNode(), # Empty, open for future use
+    routineDef.last.normalizeStmtList(map)
+  )
+
+proc normalizeImportExport(importStmt: NimNode, map: IdentMap): NimNode =
+  case importStmt.kind:
+  of nnkImportExceptStmt, nnkFromStmt, nnkExportExceptStmt:
+    importStmt.kind.newTree(importStmt[0]).add importStmt[1..^1].sortedByIt(it.strVal)
+  of nnkImportStmt, nnkExportStmt:
+    importStmt.kind.newTree(importStmt[0..^1].sortedByIt(if it.kind == nnkInfix: it.unpackInfix.left.strVal else: it.strVal)) # TODO: implemement normalizations of `import macros as m`
+  else:
+    raise newException(ValueError, $importStmt & "is not a valid import or export stmt")
+
+proc normalizeStmtList*(code: NimNode, map: var IdentMap): NimNode =
+  code.expectKind nnkStmtList
+  var normalizedTree = nnkStmtList.newTree
+
+  for index, statement in code:
+    normalizedTree.add case statement.kind:
+    of nnkCommentStmt: continue
+    of nnkVarSection..nnkConstSection:
+      var newDefSection = statement.kind.newTree()
+      for def in statement:
+        newDefSection.add def.normalizeIdentDef(map)
+      newDefSection
+    of nnkTypeSection: # TODO: Implement type normalizaation
+      statement
+    of RoutineNodes - {nnkLambda, nnkDo}: # We aren't supporting `do:` as of yet
+      statement.normalizeRoutineDef(map)
+    of nnkCallKinds:
+      statement.normalizeCall(map)
+    of nnkImportStmt, nnkFromStmt, nnkImportExceptStmt:
+      statement.normalizeImportExport(map)
+    of nnkDiscardStmt:
+      nnkDiscardStmt.newNimNode.add statement[0].normalizeValue(map)
+    else:
+      statement
+
+  normalizedTree
+
+proc createRepresentation*(fileName: string): (NimNode, IdentMap) =
+  var map: IdentMap
+  let code = parseStmt fileName.staticRead
+  (code.normalizeStmtList(map), map)
+
+when isMainModule:
+  import json
+  static:
+    const path {.strdefine.} = "../../Exercism/nim/two-fer/two_fer.nim" ##\
+    ## The path to the file to create representation for
+    ## Can be invoked with -d:path=<PATH-TO-FILE>
+    let (tree, map) = createRepresentation path
+    echo (%*{"map": map, "tree": tree.repr}).pretty
```

Comment:
Due to the `distinct string` change, this line causes `nim c src/representer.nim` to error with
```
Error: type mismatch: got <OrderedKeyValuePair[mapping.NormalizedIdent, system.string]>
but expected one of: 
[...]
proc `%`[T](table: Table[string, T] | OrderedTable[string, T]): JsonNode
  first type mismatch at position: 1
  required type for table: Table[system.string, %.T] or OrderedTable[system.string, %.T]
  but expression 'elem' is of type: OrderedKeyValuePair[mapping.NormalizedIdent, system.string]
```

Reply:
Fixed in [5d4ba6e](https://github.com/exercism/nim-representer/pull/3/commits/5d4ba6ee0afbf94274c24ab603567d4181735b9b)